Jetty 12: `ContextHandler.getTempDirectory()` does not respect the `Context.getTempDirectory()` contract

lorban opened the pull request:
The `Context.getTempDirectory()` contract states that null should never be returned, while the `ContextHandler.getTempDirectory()` implementation may return null depending on the `MultipartConfigElement`.

This PR makes sure that `ContextHandler.getTempDirectory()` never returns null.

Fixes #11396

lorban committed:
#11396 fix ContextHandler.getTempDirectory() so it never returns null as the contract mandates

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime reviewed (changes requested):
@lorban the `ServletContextHandler` now needs to be updated to make a unique directory within the Server temp directory.

A test case would be ...

* A Server with a default temp directory.
* A `ServletContextHandler` on context-path `/foo` (no temp directory defined at this level)
* A `ServletContextHandler` on context-path `/bar` (no temp directory defined at this level)
* Submit a `multipart/form-data` to `/foo/` that has 2 parts (part `"name"` and part `"file"` which causes a Temp directory write)
* Submit a `multipart/form-data` to `/bar/` that has 2 parts (part `"name"` and part `"file"` with different file contents,  which also causes a Temp directory write)

Those two usages should not step on each others toes.
The `ServletContext.TEMPDIR` attribute should also report as different for each of those `ServletContextHandler` instances.

lorban committed:
#11396 use separate tmp folder per context

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11396 fix ServletContext.TEMPDIR attribute

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11396 save created temp directory

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@joakime I think I modified the PR to handle the cases you specified.

joakime commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -290,13 +287,6 @@ public ServletContextHandler(String contextPath, SessionHandler sessionHandler,
-    public void setTempDirectory(File tempDirectory)
-    {
-        super.setTempDirectory(tempDirectory);
-        setAttribute(TEMPDIR, getTempDirectory());
```
Where does this spec attribute get set now?

lorban replied:
I've mistakenly moved that attribute from context to request. Let me move it back to where it belongs.

joakime commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextRequest.java:
```diff
@@ -304,6 +305,9 @@ public Charset getQueryEncoding()
     @Override
     public Object getAttribute(String name)
     {
+        if (ServletContext.TEMPDIR.equals(name))
+            return getContext().getTempDirectory();
```
This behaves kinda like a synthetic attribute now.

Note: the `ServletContext.TEMPDIR` is a ServletContext attribute, not a Request attribute (I think this code is in the wrong place)

Will this attribute show up when a user iterates the attributes now?

lorban replied:
You're right, let me fix that.

joakime commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/MultiPartServletTest.java:
```diff
@@ -371,6 +391,152 @@ protected void service(HttpServletRequest request, HttpServletResponse response1
+            @Override
+            protected void service(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException
+            {
+                File tmpDirFromAttribute = (File)request.getAttribute(ServletContext.TEMPDIR);
```
Using `request.getAttribute(ServletContext.TEMPDIR)` should return null.
That attribute doesn't exist on the request, it only exists on the context.

This code should be using `getServletContext().getAttribute(ServletContext.TEMPDIR)` instead.

lorban replied:
You're right, let me fix that.

joakime reviewed (changes requested):
Good progress, but the TEMPDIR attribute is in the wrong place now.  
It should only exists on the ServletContext attributes, not the HttpServletRequest attributes

filiphr commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -1124,9 +1126,15 @@ public Resource getBaseResource()
+                    tempDirectory = new File(getServer().getContext().getTempDirectory(), _contextPath);
+                    tempDirectory.mkdirs();
+                }
+            }
```
How does this behave in a concurrent scenario? From what I can see there is now a private field on the `ScopedContext` which gets assigned on the `ScopedContext` when the `getTempDirectory` method is called.  I don't know much about how all of this gets called. However, from what I could see in the code it gets called from `ServletApiRequest#getParts` and I can imagine that this can get called from multiple threads.

lorban replied:
You're correct, this code could be called by multiple concurrent threads. The current version is safe by accident, but it should be improved. Let me work on this.

lorban committed:
#11396 move ServletContext.TEMPDIR as a context attribute + cleanup test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11396 choose and create temp directory during context start

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

filiphr commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextHandler.java:
```diff
@@ -1988,6 +1979,22 @@ public String toString()
+            attributeNameSet.add(ServletContext.TEMPDIR);
+            return Collections.unmodifiableSet(attributeNameSet);
+        }
+
```
Isn't the previous code with the `setAttribute(TEMPDIR, getTempDirectory());` in the `ServletContextHandler#setTempDirectory` a bit more performant than this one? 

I don't know the code base much, but from what I could see if the `TEMPDIR` attribute is set in the `ServletContextHandler#setTempDir` you won't need to create this copy of the existing set.

lorban replied:
You're right, with my latest changes this special getAttribute() implementation isn't needed anymore.

lorban committed:
#11396 restore previous setTempDirectory impl

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11396 fix tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add debugging info

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

filiphr commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -669,8 +680,33 @@ protected void createTempDirectory()
                     throw new IllegalArgumentException("Failed to delete temp dir: " + tempDirectory);
-                if (!tempDirectory.mkdirs())
-                    throw new IllegalArgumentException("Unable to create temp dir: " + tempDirectory);
+                try
+                {
+                    Files.createDirectories(tempDirectory.toPath());
+                }
+                catch (IOException e)
+                {
+                    Path path = tempDirectory.toPath();
+                    while (!"/".equals(path.toString()))
+                    {
+                        System.out.println("Path '" + path + "' exists? " + Files.exists(path));
+                        if (Files.exists(path))
+                        {
+                            try
+                            {
+                                UserPrincipal owner = Files.getOwner(path);
+                                Set<PosixFilePermission> posixFilePermissions = Files.getPosixFilePermissions(path);
+                                System.out.println("  user name: " + System.getProperty("user.name") + " file owner: " + owner + " perm: " + posixFilePermissions);
+                            }
+                            catch (Exception ex)
+                            {
+                                ex.printStackTrace();
+                            }
+                        }
+                        path = path.getParent();
+                    }
+                    throw new IllegalArgumentException("Unable to create temp dir: " + tempDirectory, e);
```
I see that you are using `System.out` here. Doesn't it make more sense to use the logger? I was also comparing this a bit with the way you are creating the temp directory in `WebInfConfiguration`

https://github.com/jetty/jetty.project/blob/55e307224782b4b7ed4cc0756deafdd492c9417c/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebInfConfiguration.java#L155-L182

A slightly different approach is used there. e.g. `Files.createTempDirectory` from Java itself. 

Again, I have no clue about the codebase, and I'm just comparing with a subset of things I can see.

lorban replied:
That was very temporary to help me troubleshoot some test failures in CI. Now that I've figured them out, I reverted that ugly change.

filiphr replied:
oops sorry for the noise in that case. I won't mix in your review anymore. Thank you very much for looking into my issue and working on a fix for it.

lorban committed:
#11396 add debugging info

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11396 rename variable for readability

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#11396 fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12/11396-incorrect-ContextHandler-getTempDirectory

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12/11396-incorrect-ContextHandler-getTempDirectory

lorban commented:
@gregw @janbartel I need your wisdom here.

This PR started as something simple: the `org.eclipse.jetty.server.Context.getTempDirectory()` contract says: _return a non-null temporary directory_ and the `ContextHandler.ScopedContext` implementation does not respect that contract as it returns null when no temp directory is explicitly configured, which makes multipart servlets throw NPE as they make use of the temp directory internally.

So I changed the `ContextHandler.ScopedContext.getTempDirectory()` implementation to never return null by delegating the call to `getServer().getContext().getTempDirectory()` when no temp directory is explicitly configured. That looked easy, but that was, unfortunately, the beginning of a rabbit's hole.

But that did not work as each context should have its own temp directory, which may be persistent, or there may be different running Jetty servers with the same server temp directory (which defaults to `java.io.tmpdir`), so the contexts' temp folder names must be unique (we *do* run into that case simply by running our unit tests in parallel), then @joakime reminded me that we generate temp folder names using a well-known format for webapps (using `WebInfConfiguration.getCanonicalNameForWebAppTmpDir()`) so we may want to do something similar for core contexts. And now, every context has a temp directory that must be created, deleted, checked for persistence and so on while that is rarely needed.

So that got me wondering: is my fix over-engineered? Is the `org.eclipse.jetty.server.Context.getTempDirectory()` too strict in forbidding to return null? Shouldn't multipart servlets be able to cope with that lack of temp folder?

lorban committed:
#11396 make more familiar temp dir name + correctly handle persistent setting

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented:
> But that did not work as each context should have its own temp directory

I still feel that every `ServletContextHandler` should have it's own private temp directory (as the javadoc for `ServletContext` says).
But having every `ContextHandler` be its own unique temp directory feels like overkill.

janbartel commented:
@lorban @joakime please see my comment on the original issue.

The only change we need to make is to ensure that `ScopedContext` falls back to `java.io.tmpdir` after trying the `Server.getTempDir()`.

`ServletContextHandler`s in all previous versions of jetty have never had their own, unique temporary directory - that is something we set up for `WebAppContext`s. It's a separate issue and PR if you now want to push the `WebAppContext` temp dir behaviour back down into `ServletContextHandler`s or indeed `ContextHandler`s.  

Let's fix the immediate problem with multipart by making `ScopedContext` honour the javadoc that says it should never return null.

joakime commented:
> @lorban @joakime please see my comment on the original issue.
> 
> The only change we need to make is to ensure that `ScopedContext` falls back to `java.io.tmpdir` after trying the `Server.getTempDir()`.
> 
> `ServletContextHandler`s in all previous versions of jetty have never had their own, unique temporary directory - that is something we set up for `WebAppContext`s. It's a separate issue and PR if you now want to push the `WebAppContext` temp dir behaviour back down into `ServletContextHandler`s or indeed `ContextHandler`s.
> 
> Let's fix the immediate problem with multipart by making `ScopedContext` honour the javadoc that says it should never return null.

So rollback to commit https://github.com/jetty/jetty.project/pull/11397/commits/f1e7e8d7798fb0458007d1fe2a3ce822c3a30c73 ?

janbartel commented:
@joakime yes, I guess so.

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12/11396-incorrect-ContextHandler-getTempDirectory

lorban committed:
Revert "#11396 fix test"

This reverts commit 353faf986e9d5063852f0be274bd06c30c54ee4e.

lorban committed:
Revert "#11396 make more familiar temp dir name + correctly handle persistent setting"

This reverts commit 64f2e8a87b332e107c78ab23a1f431d27295b53b.

lorban committed:
Revert "#11396 add debugging info"

This reverts commit 06a36bed95b9a74065197e614d964437e2643d8e.

lorban committed:
Revert "#11396 rename variable for readability"

This reverts commit 418e1b60effc0c259d12654d217e7e505428b5ab.

lorban committed:
Revert "#11396 fix tests"

This reverts commit 07bacca9380e60699b5990b72a2d80ac222cae59.

lorban committed:
Revert "add debugging info"

This reverts commit 3395f910f4bad4058980666aa560c1a182f234db.

lorban committed:
Revert "#11396 restore previous setTempDirectory impl"

This reverts commit 9ed0a58c5c2032c4e216ec506653732fa20b220b.

lorban committed:
Revert "#11396 choose and create temp directory during context start"

This reverts commit a0811b1cb08e715515656d503802b071d0768178.

lorban committed:
Revert "#11396 move ServletContext.TEMPDIR as a context attribute + cleanup test"

This reverts commit eeafe811b41f8a39e2edea2fe0cc4372d66490dc.

lorban committed:
Revert "#11396 save created temp directory"

This reverts commit 7ba67ce7b9249af99c33463e6b1bd2af3c5df0cd.

lorban committed:
Revert "#11396 fix ServletContext.TEMPDIR attribute"

This reverts commit 2e5250c4db510d19bc9043e6f81bbbdba0407c6d.

lorban committed:
Revert "#11396 use separate tmp folder per context"

This reverts commit 4c9ee9a3a6597997202586134ce8eea62c7fd3c8.

lorban commented:
@joakime nudge
